### PR TITLE
Avoid project read lock in SDK version calculation

### PR DESCRIPTION
**Pull request by drewnoakes**
Use `TryGetCurrentConfigurationGeneralPropertiesSnapshot` rather than `GetConfigurationGeneralPropertiesAsync` as the latter will take a project read lock, while the former uses the most recent available snapshot. For the SDK version, this is unlikely to change throughout the lifetime of the project, so even an older version of this value will be fine.

This code runs for every project during solution load.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/project-system/pull/9347)

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Telemetry/SDKVersionTelemetryServiceComponent.SDKVersionTelemetryServiceInstance.cs [hunk 1]
```diff
@@ -29,25 +29,35 @@ internal class SDKVersionTelemetryServiceInstance : OnceInitializedOnceDisposed,
...
                 _projectVsServices.ThreadingService.RunAndForget(async () =>
                 {
-                    // Wait for the project to be loaded so that we don't prematurely load the active configuration
+                    // Wait for the project to be loaded so that we don't prematurely load the active configuration.
                     await _unconfiguredProjectTasksService.ProjectLoadedInHost;
 
-                    ConfigurationGeneral projectProperties = await _projectVsServices.ActiveConfiguredProjectProperties.GetConfigurationGeneralPropertiesAsync();
+                    // We use TryGetCurrentConfigurationGeneralPropertiesSnapshot rather than GetConfigurationGeneralPropertiesAsync
+                    // as the latter will take a project read lock, while the former uses the most recent available
+                    // snapshot. For the SDK version, this is unlikely to change throughout the lifetime of the project,
```
Comment (adamint) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Telemetry/SDKVersionTelemetryServiceComponent.SDKVersionTelemetryServiceInstance.cs:40]: could you mention the circumstances under which the most recent snapshot will change?

**Review by adamint — APPROVED** [review]

**Comment by adamint** [comment]
Are the test failures related to this change?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Telemetry/SDKVersionTelemetryServiceComponent.SDKVersionTelemetryServiceInstance.cs (hunk 1 above)
Comment (lifengl) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Telemetry/SDKVersionTelemetryServiceComponent.SDKVersionTelemetryServiceInstance.cs:40]: I feel the logic here is to collect a snapshot of the SDK version of the project in the telemetry, no matter when it is captured. Actually, the code here is only to ensure this code to run after a project is fully loaded once, but nothing to ensure when it must be completed, so there is no meaningful 'current' state of the project, no matter which API is being used.

**Review by lifengl — APPROVED** [review]

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Telemetry/SDKVersionTelemetryServiceComponent.SDKVersionTelemetryServiceInstance.cs (hunk 1 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Telemetry/SDKVersionTelemetryServiceComponent.SDKVersionTelemetryServiceInstance.cs]: @adamint there are, generally, two types of data you might want:

1. The value of the project at the current moment.
2. The value of the snapshot at the current moment.

These sound the same, but the snapshot is the result of evaluation/build, and those take time. If you are interested in only the snapshot, then getting those values is just a matter of looking up the latest immutable snapshot and returning it, which is easy. In this case we wait for the proejct to load, so the snapshot is guaranteed to be there. The `Debug.Assert` a bit lower down is trying to catch if that assumption is ever broken.

In contrast, getting the current project value requires some synchronisation between the project XML and the outputs of evaluation/build. In some places we use `GetLatestValueAsync` on a `ProjectValueDataSource` to await a value that might be flowing through the dataflow graph (using project value version). @lifengl explained, when reporting this issue, that the previous method actually takes a project read lock, which not only waits but also excludes other actions.

Using a snapshot is lighter and non-blocking. In our case, we do not expect the SDK version to change over time, so even "old" values are fine here.

**Comment by drewnoakes** [comment]
Getting the tests passing required some refactoring. The new use of `TryGetCurrentConfigurationGeneralPropertiesSnapshot` is incredibly painful to mock. Furthermore, the old `SDKVersionTelemetryServiceComponent` was implemented as a multi-life component on a non-dynamic component.

I addressed this by having only a single lifetime for the component, pulling the acquisition of the SDK version into a virtual method that can be mocked, and flattening two types into one (renaming it and the corresponding unit test in the process).

Also uses an xUnit theory to merge all tests into one. There was so much churn that I then converted to file-scoped namespaces.

Each of these in their own commit.

**Pull request merged**